### PR TITLE
Fix `noop_update_total` is not being updated when using the `_bulk`

### LINKED ISSUE #105742: `noop_update_total` is not being updated when using the `_bulk`
### Elasticsearch Version

8.12.1

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

official image

### Problem Description

`noop_update_total` is not being updated when using the `_bulk`, but it is being updated when using the `_update`.

### Steps to Reproduce

Bulk
```
DELETE noop_update_demo
// index
POST noop_update_demo/_bulk
{"create":{"_id":"1"}}
{"foo":"bar"}
{"create":{"_id":"2"}}
{"foo":"bar"}

// noop update
POST noop_update_demo/_bulk
{"update":{"_id":"1"}}
{"doc":{"foo": "bar"}}
{"update":{"_id":"2"}}
{"doc":{"foo": "bar"}}

// index_total = 2, noop_update_total = 0 (should be 2)
GET _stats?filter_path=indices.noop_update_demo.total.indexing
```
Return:
```json
{
  "indices": {
    "noop_update_demo": {
      "total": {
        "indexing": {
          "index_total": 2,
          "index_time_in_millis": 1,
          "index_current": 0,
          "index_failed": 0,
          "delete_total": 0,
          "delete_time_in_millis": 0,
          "delete_current": 0,
          "noop_update_total": 0,  //Should be 2
          "is_throttled": false,
          "throttle_time_in_millis": 0,
          "write_load": 0.0002581126849195685
        }
      }
    }
  }
}
```

Update
```
DELETE noop_update_demo
// index
POST noop_update_demo/_bulk
{"create":{"_id":"1"}}
{"foo":"bar"}
{"create":{"_id":"2"}}
{"foo":"bar"}

// noop update
POST noop_update_demo/_update/1
{"doc":{"foo": "bar"}}
POST noop_update_demo/_update/2
{"doc":{"foo": "bar"}}

// index_total = 2, noop_update_total = 2
GET _stats?filter_path=indices.noop_update_demo.total.indexing
```

Return:
```json
{
  "indices": {
    "noop_update_demo": {
      "total": {
        "indexing": {
          "index_total": 2,
          "index_time_in_millis": 1,
          "index_current": 0,
          "index_failed": 0,
          "delete_total": 0,
          "delete_time_in_millis": 0,
          "delete_current": 0,
          "noop_update_total": 2,  //
          "is_throttled": false,
          "throttle_time_in_millis": 0,
          "write_load": 0.0002900231347040275
        }
      }
    }
  }
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by amooncake**
Closes #105742

### CODE REVIEW

**Comment by arteam** [comment]
buildkite test this

**Comment by amooncake** [comment]
@arteam I fixed the bwc test cases based on the build title, but I don’t have access to buildkite, is there any way to know which test cases failed?

**Comment by arteam** [comment]
Hi @amooncake!

First, sorry for a late response and secondly, thank you very for working on this issue and your interest in Elasticsearch! 
<details>
    <summary>The buildkite failure was</summary>

```
> Task :qa:smoke-test-multinode:yamlRestTest
REPRODUCE WITH: ./gradlew ':qa:smoke-test-multinode:yamlRestTest' --tests "org.elasticsearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT.test {yaml=bulk/12_noop/Noop}" -Dtests.seed=C95621318FC9D8DB -Dtests.locale=es-AR -Dtests.timezone=Antarctica/Mawson -Druntime.java=22
org.elasticsearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT > test {yaml=bulk/12_noop/Noop} FAILED
    java.lang.AssertionError: Failure at [bulk/12_noop:31]:
    Expected: <2>
         but: was <4>
        at __randomizedtesting.SeedInfo.seed([C95621318FC9D8DB:41021EEB2135B523]:0)
        at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:593)
        at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:536)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
        at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster$1.evaluate(DefaultLocalElasticsearchCluster.java:47)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
        at java.base/java.lang.Thread.run(Thread.java:1570)
        Caused by:
        java.lang.AssertionError:
        Expected: <2>
             but: was <4>
            at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
            at org.junit.Assert.assertThat(Assert.java:964)
            at org.junit.Assert.assertThat(Assert.java:930)
            at org.elasticsearch.test.rest.yaml.section.MatchAssertion.doAssert(MatchAssertion.java:99)
            at org.elasticsearch.test.rest.yaml.section.Assertion.execute(Assertion.java:65)
            at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:573)
            ... 41 more
```
  
</details>

but I see that you've already fixed this by adding the version check :+1: . Let me sync this branch with main and run the test suite in  buildkite.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
buildkite test this

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            number_of_shards: 1
+  - do:
+      bulk:
+        refresh: true
+        body: |
+          { "create": { "_index": "test_1", "_id": "1"} }
+          { "foo": "bar" }
+          { "create": { "_index": "test_1", "_id": "2"} }
+          { "foo": "bar" }
+  - do:
```
Comment (arteam) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml:20]: I think it makes sense to check the index stats after the first bulk requests to make sure that it was not a no-op update
```
  - do:
      indices.stats: { index: test_1 }

  - match: { _all.total.indexing.index_total: 2 }
  - match: { _all.total.indexing.noop_update_total: 0 }
```

**Review by arteam — COMMENTED** [review]
The change looks good! I've left one suggestion to improve the test, but overall the PR is ready to be merged.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml (hunk 1 above)
Comment (amooncake) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml:20]: Thanks for the review! I've made the suggested changes.

**Comment by nik9000** [comment]
> I've run the suite multiple times locally with different seeds and it seems that the value `_all.total.indexing.index_total` is non-deterministic in an environment where we run multiple tests. Do you think it would make sense to remove `index_total` assertions and only check `noop_update_total`?

Could we just check the total for the index you make?

**Comment by amooncake** [comment]
@arteam @nik9000 After multiple tests, I found the non-deterministic `index_total` due to varying index statuses(default 1 replica, green:4, yellow:2). Adding status checks resolved the issue.

The test cases also have been changed to only check the target index.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
buildkite test this

**Comment by arteam** [comment]
@amooncake Thank you very much for iterating on this PR! 

Indeed, waiting for the cluster to be green  seems to be making `index_total` asserts deterministic. Unfortunately, it looks like adding an additional replica makes the test suite fail with a 408 HTTP error since the replica doesn't get allocated in time. 

```
java.lang.AssertionError: Failure at [bulk/12_noop:13]: expected [2xx] status code but api [cluster.health] returned [408 Request Timeout] [{"cluster_name":"test-cluster","status":"yellow","timed_out":true,"number_of_nodes":1,"number_of_data_nodes":1,"active_primary_shards":1,"active_shards":1,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":50.0}]
```

What do you think about keeping the number of replicas to 0 and adjusting the corresponding `index_total` match assertions?

```diff
===================================================================
diff --git a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml
--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml	(revision 2051a83912fefee2b216f2eb08dbc27413e31755)
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/bulk/12_noop.yml	(date 1711618151996)
@@ -9,7 +9,7 @@
         body:
           settings:
             number_of_shards: 1
-            number_of_replicas: 1
+            number_of_replicas: 0
   - do:
       cluster.health:
         wait_for_status: green
@@ -24,7 +24,7 @@
   - do:
       indices.stats: { index: test_1 }
 
-  - match: { indices.test_1.total.indexing.index_total: 4 }
+  - match: { indices.test_1.total.indexing.index_total: 2 }
   - match: { indices.test_1.primaries.indexing.index_total: 2 }
   - match: { indices.test_1.total.indexing.noop_update_total: 0 }
   - match: { indices.test_1.primaries.indexing.noop_update_total: 0 }
@@ -43,7 +43,7 @@
   - do:
       indices.stats: { index: test_1 }
 
-  - match: { indices.test_1.total.indexing.index_total: 4 }
+  - match: { indices.test_1.total.indexing.index_total: 2 }
   - match: { indices.test_1.primaries.indexing.index_total: 2 }
   - match: { indices.test_1.total.indexing.noop_update_total: 2 } # total noop == primaries noop
   - match: { indices.test_1.primaries.indexing.noop_update_total: 2 }

```

**Comment by amooncake** [comment]
> @amooncake Thank you very much for iterating on this PR!
> 
> Indeed, waiting for the cluster to be green seems to be making `index_total` asserts deterministic. Unfortunately, it looks like adding an additional replica makes the test suite fail with a 408 HTTP error since the replica doesn't get allocated in time.
> 
> ```
> java.lang.AssertionError: Failure at [bulk/12_noop:13]: expected [2xx] status code but api [cluster.health] returned [408 Request Timeout] [{"cluster_name":"test-cluster","status":"yellow","timed_out":true,"number_of_nodes":1,"number_of_data_nodes":1,"active_primary_shards":1,"active_shards":1,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":50.0}]
> ```
> 
> What do you think about keeping the number of replicas to 0 and adjusting the corresponding `index_total` match assertions?
> 
@arteam Thanks for the test case information and modification suggestions!
I initially suspected performance issues were causing the index's yellow state. However, it turns out there are different number of data nodes. As waiting for a green state wouldn't be helpful here, I've set the replica shards to 0 to address this.

**Review by arteam — APPROVED** [review]

**Comment by arteam** [comment]
@amooncake Thank you for your contribution!

**Pull request merged**